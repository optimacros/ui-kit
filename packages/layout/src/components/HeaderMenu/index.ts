export * from './HeaderMenu';
export * from './type';
