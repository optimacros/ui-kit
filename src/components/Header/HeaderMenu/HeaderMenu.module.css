.Container {
  max-height: 200px;
  position: relative;
  z-index: 1000;
  flex: 1;
  width: 100%;
}

.Container_special {
  & .Menu {
    padding: 0;
  }

  & .MenuItem {
    &:last-child {
      & .Element {
        padding-right: 16px;
      }
    }
  }

  & .Element_Title {
    font-size: 14px;
    max-width: 180px;
  }
}

/* -------------------------------------------------- */
/* Menu */
/* -------------------------------------------------- */
.Menu {
  position: relative;
  display: flex;
  flex-direction: row;
  margin: 0;
  padding: 2px 0 0;
  list-style-type: none;

  & > li {
    position: relative;
  }
}

/* -------------------------------------------------- */
/* Submenu */
/* -------------------------------------------------- */
.SubMenu {
  position: absolute;
  left: -1px;
  z-index: 10;
  margin: 0;
  padding: 5px 0;
  border: 1px solid var(--headerMenu-subMenu-border);
  border-top: none;
  box-shadow: 6px 8px 8px 0 var(--headerMenu-subMenu-shadow);
  background: var(--headerMenu-subMenu-bg);

  & .SubMenu {
    position: fixed;
    margin: -8px 0 0 0;
    border-top: 1px solid var(--color-scale-gray-8);
  }
}

/*
Attribute max-width: 250px; is the correction of an error in the task :
https://optimacros.myjetbrains.com/youtrack/issue/OM-1575
*/

.SubMenuScrollList {
  list-style-type: none;
  margin: 0;
  padding: 0;
  max-height: 420px;
  max-width: 250px;
  overflow-x: hidden;
  overflow-y: auto;
}

/* -------------------------------------------------- */
/* Menu Item */
/* -------------------------------------------------- */
.MenuItem {
  position: static;
  padding: 0;
  cursor: pointer;
  white-space: nowrap;
  border: 1px solid transparent;
  border-bottom-width: 2px;
  font-size: 1.3rem;
  margin: 0 2px;

  &__disabled {
    cursor: auto;

    & > * {
      opacity: 0.4;
    }
  }

  &_parent {
    &:not(:first-child) {
      border-top: 1px solid var(--color-scale-gray-8) !important;
    }
  }

  &_child {
    padding-left: 10px !important;
  }

  & .Element_IconContainer {
    margin-right: 12px;
  }

  & .Element_Icon {
    font-size: 24px;
    color: #fff;
  }

  &:not(.MenuItem_disabled):hover {
    background: var(--headerMenu-subMenu-menuItem-hover);
  }

  &:hover > .SubMenu {
    display: block;
  }

  &:first-child,
  &:last-child {
    margin: 0;
  }
}

.SubMenu {
  & .MenuItem {
    color: var(--primary-text-color);
    margin: 2px 0;
    border: none;

    &:not(.MenuItem_disabled):hover {
      border: none;
      box-shadow: none;
      background: var(--headerMenu-subMenu-menuItem-hover);
      color: var(--primary-text-color);
    }
  }

  & .Element_IconContainer {
    margin-right: 0;
  }

  & .Element_Icon {
    font-size: 15px;
    color: var(--primary-text-color);

    svg {
      max-width: 15px;
      max-height: 15px;
    }
  }
}

/* -------------------------------------------------- */
/* Element */
/* -------------------------------------------------- */
.ElementContainer {
  padding: 4px 8px 8px;
  display: flex;
  align-items: center;
}

.Element {
  min-height: 23px;
  padding: 0 8px;
  display: flex;
  align-items: center;

  &_IconContainer {
    width: auto;
    text-align: center;
  }

  &_Icon {
    display: flex;
    align-items: center;
    justify-content: center;

    & img {
      max-width: 15px;
      max-height: 15px;
    }
  }

  &_Title {
    flex: 1;
    text-overflow: ellipsis;
    overflow: hidden;
  }

  &_Arrow {
    font-size: 15px;
    color: var(--primary-text-color);
  }
}

.SubMenu {
  & .Element {
    color: var(--primary-text-color);
    padding: 8px 4px 8px 32px;

    &_withIcon {
      padding-left: 0;
    }

    &_IconContainer {
      width: 32px;
    }

    &_Title {
      margin-right: 20px;
      text-overflow: ellipsis;
      overflow: hidden;
    }
  }
}

.Search {
  position: relative;
  border: 1px solid var(--light-divider-color);
  margin: 4px 8px;

  &:hover {
    border-color: #d3d3d3;
  }

  & .inputElement {
    padding: 8px;
    padding-right: 32px;
    border: none;
    font-size: 13px;

    &:focus:not([disabled]):not([readonly]) + .icon {
      color: #c1c1c1;
    }
  }

  & .hint {
    font-size: 13px;
    left: 48px;
  }
}

.ListSearchClear {
  position: absolute;
  display: flex !important;
  align-items: center;
  justify-content: center;
  top: 0;
  right: 0;
  color: #c1c1c1;
  cursor: pointer;
  font-size: 20px;
  height: 100%;
  width: 36px;

  &:hover {
    color: #606060;
  }
}

.HiddenPlaceholder {
  height: 0;
  visibility: hidden;
  background: lightblue;
  margin-right: 40px;
}
