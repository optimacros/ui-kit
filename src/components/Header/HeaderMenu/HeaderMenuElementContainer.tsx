// @ts-nocheck
import classNames from 'classnames'
import _ from 'lodash'
import { observer } from 'mobx-react'
import React from 'react'
import { Icon } from 'ui-kit-core'

import styles from './HeaderMenu.module.css'

interface Props {
	isFirstLevel: boolean;
	element: Element;
	rootElementNode: Node;
	elements: Array<Element>;
}

@observer
export default class HeaderMenuElementContainer extends React.Component<Props> {
    render() {
        const { element } = this.props

        const className = classNames({
            [styles.Element]: true,
            [styles.Element_withIcon]: !!element.icon,
            [styles.ElementContainer]: true,
        })

        return (
            <div
                className={className}
                disabled={element.disabled}
                onClick={this._onClick}
            >
                {this.renderIcon(element)}

                <div className={styles.Element_Title}>{element.title}</div>

                {this.renderArrowIcon()}
            </div>
        )
    }

    renderIcon(element) {
        if (!element.icon) {
            return null
        }

        return (
            <div className={styles.Element_IconContainer}>
                <Icon
                    className={styles.Element_Icon}
                    value={element.icon}
                />
            </div>
        )
    }

    renderArrowIcon() {
        const { element, isFirstLevel } = this.props

        if (isFirstLevel || _.isEmpty(element.children) || element.disabled) {
            return null
        }

        return (
            <div>
                <Icon
                    className={styles.Element_Arrow}
                    value="navigate_next"
                />
            </div>
        )
    }

    _onClick = () => {
        const { element } = this.props

        if (!element.disabled && element.open) {
            element.open()
        }
    }
}
