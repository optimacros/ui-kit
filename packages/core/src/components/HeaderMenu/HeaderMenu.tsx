import classNames from 'classnames';
import { isEmpty, map } from 'lodash';
import type { FC, JSX } from 'react';

import { HeaderMenuElement } from './HeaderMenuElement';
import type { MenuElement } from './type';

import styles from './HeaderMenu.module.css';

export interface HeaderMenuProps {
    elements: MenuElement[];
    className?: string;
}

export const HeaderMenu: FC<HeaderMenuProps> = (props) => {
    const { elements, className } = props;

    if (isEmpty(elements)) {
        return null;
    }

    const finalClassName = classNames(styles.Container, className);

    const renderList = (): (JSX.Element | null)[] => {
        return map(elements, (element) => {
            if (element.hidden) {
                return null;
            }

            return <HeaderMenuElement key={element.id} element={element} firstLevel />;
        });
    };

    return (
        <div className={finalClassName}>
            <ul className={styles.Menu}>{renderList()}</ul>
        </div>
    );
};
