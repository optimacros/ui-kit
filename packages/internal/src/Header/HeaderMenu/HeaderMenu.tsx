import * as _ from '@optimacros-ui/utils';
import { observer } from 'mobx-react';
import { Component } from 'react';

import { HeaderMenuElement } from './HeaderMenuElement';

import styles from './HeaderMenu.module.css';

@observer
export class HeaderMenu extends Component<{
    elements: Array<any>;
}> {
    render() {
        if (_.isEmpty(this.props.elements)) {
            return null;
        }

        return (
            <div className={styles.Container}>
                <ul className={styles.Menu}>{this.renderList()}</ul>
            </div>
        );
    }

    renderList() {
        return _.map(this.props.elements, (element) => {
            if (element.hidden) {
                return null;
            }

            const key = element.id || element.entityLongId || element.title;

            return <HeaderMenuElement key={key} element={element} firstLevel />;
        });
    }
}
