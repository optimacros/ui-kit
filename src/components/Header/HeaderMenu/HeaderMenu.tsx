// @ts-nocheck
import _ from 'lodash'
import { observer } from 'mobx-react'
import PropTypes from 'prop-types'
import React, { Component } from 'react'

import { HeaderMenuElement } from './HeaderMenuElement'

import styles from './HeaderMenu.module.css'

@observer
export class HeaderMenu extends Component {
    static propTypes = {
        elements: PropTypes.array.isRequired,
    }

    render() {
        if (_.isEmpty(this.props.elements)) {
            return null
        }

        return (
            <div className={styles.Container}>
                <ul className={styles.Menu}>{this.renderList()}</ul>
            </div>
        )
    }

    renderList() {
        return _.map(this.props.elements, element => {
            if (element.hidden) {
                return null
            }

            const key = element.id || element.entityLongId || element.title

            return (
                <HeaderMenuElement
                    key={key}
                    element={element}
                    firstLevel
                />
            )
        })
    }
}
