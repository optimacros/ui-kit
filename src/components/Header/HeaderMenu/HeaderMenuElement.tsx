// @ts-nocheck
import classNames from 'classnames'
import _ from 'lodash'
import { observer } from 'mobx-react'
import React from 'react'

import HeaderMenuElementContainer from './HeaderMenuElementContainer'
import HeaderSubMenu from './HeaderSubMenu'

import styles from './HeaderMenu.module.css'

interface Props {
    firstLevel: boolean;
    element: Element;
    rootElementNode: Node;
    elements: Array<Element>;
}

@observer
export class HeaderMenuElement extends React.Component<Props> {
    constructor(props) {
        super(props)

        this._node = React.createRef()
    }

    state = {
        showMenu: false,
    }

    render() {
        const { element } = this.props

        if (element.hidden) {
            return null
        }

        const className = classNames({
            [styles.MenuItem]: true,
            [styles.MenuItem__disabled]: element.disabled,
            [styles.MenuItem_child]: element.isChild,
            [styles.MenuItem_parent]: element.isParent,
        })

        return (
            <li
                ref={this._node}
                className={className}
                title={element.tooltip || element.title}
                onMouseEnter={this._onMouseEnter}
                onMouseLeave={this._onMouseLeave}
            >
                <HeaderMenuElementContainer
                    element={element}
                    isFirstLevel={this.props.firstLevel}
                />
                {this.renderSubMenu()}
            </li>
        )
    }

    renderSubMenu() {
        const { element } = this.props

        if (!this.state.showMenu || _.isEmpty(element.children)) {
            return null
        }

        return (
            <HeaderSubMenu
                element={element}
                elements={element.children}
                firstLevel={this.props.firstLevel}
                rootElementNode={this._node}
            />
        )
    }

    _onMouseEnter = () => {
        this.setState(() => {
            return {
                showMenu: true,
            }
        })
    }

    _onMouseLeave = () => {
        this.setState(() => {
            return {
                showMenu: false,
            }
        })
    }
}
