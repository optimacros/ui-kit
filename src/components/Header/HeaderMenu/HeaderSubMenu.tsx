// @ts-nocheck
import _ from 'lodash';
import { observer } from 'mobx-react';
import React from 'react';
import { FontIcon, Input } from 'ui-kit-core';

import { HeaderMenuElement } from './HeaderMenuElement';

import styles from './HeaderMenu.module.css';

const HEADER_MENU_OFFSET_SUBMENU = 2;
const HEADER_MENU_OFFSET_FROM_WINDOW = 20;

interface Props {
    firstLevel: boolean;
    element: Element;
    rootElementNode: Node;
    elements: Array<Element>;
}

@observer
export default class HeaderSubMenu extends React.Component<Props> {
    constructor(props) {
        super(props);

        this._rootMenuNode = React.createRef();

        this.state = {
            searchValue: '',
        };
    }

    componentDidMount() {
        /**
         * ! После монтирования компонента в dom, устанавливаем корректную позицию, относительно родительского элемента
         */
        if (!this.props.firstLevel) {
            this._setStyles();
        }
    }

    render() {
        const { elements, element } = this.props;

        if (_.isEmpty(elements)) {
            return null;
        }

        return (
            <div
                ref={this._rootMenuNode}
                className={styles.SubMenu}
                onMouseLeave={this._onMouseLeave}
            >
                {this.renderSearch()}

                <ul className={styles.SubMenuScrollList} id={element.type}>
                    {this.renderList()}
                </ul>
            </div>
        );
    }

    renderSearch() {
        if (!this.props.element.showSearch) {
            return null;
        }

        return (
            <div className={styles.Search}>
                <div aria-hidden="true" className={styles.HiddenPlaceholder}>
                    {this.props.element.placeholder || 'Search'}
                </div>

                <Input
                    name="search"
                    placeholder={this.props.element.placeholder || 'Search'}
                    theme={styles}
                    value={this.state.searchValue}
                    collapsed
                    onChange={this.onChange}
                />

                {this.state.searchValue && (
                    <FontIcon
                        className={styles.ListSearchClear}
                        value="close"
                        onClick={this.onClearSearch}
                    />
                )}
            </div>
        );
    }

    renderList() {
        const formattedSearchValue = _.toLower(_.trim(this.state.searchValue));
        const hasSearch = formattedSearchValue !== '';

        return _.map(this.props.elements, (element) => {
            if (element.hidden) {
                return null;
            }

            if (hasSearch && !_.includes(_.toLower(element.title), formattedSearchValue)) {
                return null;
            }

            const { entityLongId, id, title } = element;
            const key = entityLongId || id || title;

            return <HeaderMenuElement key={key} element={element} />;
        });
    }

    onChange = (value) => {
        this.setState({
            searchValue: value,
        });
    };

    onClearSearch = () => {
        this.onChange('');
    };

    _setStyles = () => {
        const node = this.props.rootElementNode.current;
        const menu = this._rootMenuNode.current;

        if (menu && node) {
            const {
                top: parentTop,
                left: parentLeft,
                width: parentWidth,
            } = node.getBoundingClientRect();
            const { height: menuHeight, width: menuWidth } = menu.getBoundingClientRect();
            const top = this._getTopPosition(parentTop, menuHeight);
            const left = this._getLeftPosition(parentLeft, menuWidth, parentWidth);

            menu.style.top = `${top}px`;
            menu.style.left = `${left}px`;
        }
    };

    _getTopPosition(parentTop, menuHeight) {
        const windowsHeight = window.innerHeight;

        if (parentTop + menuHeight < windowsHeight) {
            return parentTop;
        }

        return windowsHeight - menuHeight - HEADER_MENU_OFFSET_FROM_WINDOW;
    }

    _getLeftPosition(parentLeft, menuWidth, parentWidth) {
        const canRight = parentLeft + menuWidth + parentWidth < window.innerWidth;
        const positionForRight = parentLeft + parentWidth - HEADER_MENU_OFFSET_SUBMENU;
        const positionForLeft = parentLeft - menuWidth + HEADER_MENU_OFFSET_SUBMENU;

        return canRight ? positionForRight : positionForLeft;
    }
}
