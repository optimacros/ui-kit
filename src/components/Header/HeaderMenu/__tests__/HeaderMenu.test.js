import React from 'react'
import renderer from 'react-test-renderer'

import HeaderMenu from 'common/components/HeaderMenu'

describe('HeaderMenu', function () {
    beforeEach(() => {
        this.elements = [
            {
                id: 'element 1',
                title: 'element 1',
                children: [
                    {
                        id: 'element 1.1',
                        title: 'element 1.1',
                    },
                    {
                        id: 'element 1.2',
                        title: 'element 1.2',
                    },
                    {
                        id: 'element 1.3',
                        title: 'element 1.3',
                    },
                    {
                        id: 'element 1.4',
                        title: 'element 1.4',
                    },
                ],
            },
            {
                id: 'element 2',
                title: 'element 2',
                children: [
                    {
                        id: 'element 2.1',
                        title: 'element 2.1',
                        icon: 'list',
                    },
                    {
                        id: 'element 2.2',
                        title: 'element 2.2',
                        icon: 'list',
                        children: [
                            {
                                id: 'element 2.2.1',
                                title: 'element 2.2.1',
                            },
                            {
                                id: 'element 2.2.2',
                                title: 'element 2.2.2',
                            },
                            {
                                id: 'element 2.2.3',
                                title: 'element 2.2.3',
                                icon: 'list',
                                disabled: true,
                            },
                            {
                                id: 'element 2.2.4',
                                title: 'element 2.2.4',
                            },
                        ],
                    },
                    {
                        id: 'element 2.3',
                        title: 'element 2.3',
                        icon: 'list',
                        disabled: true,
                    },
                    {
                        id: 'element 2.4',
                        title: 'element 2.4',
                        icon: 'list',
                    },
                ],
            },
            {
                id: 'element 3',
                title: 'element 3',
                children: [
                    {
                        id: 'element 3.1',
                        title: 'element 3.1',
                    },
                    {
                        id: 'element 3.2',
                        title: 'element 3.2',
                    },
                    {
                        id: 'element 3.3',
                        title: 'element 3.3',
                    },
                    {
                        id: 'element 3.4',
                        title: 'element 3.4',
                    },
                ],
            },
            {
                id: 'element 4',
                title: 'element 4',
                disabled: true,
            },
        ]
    })

    it('simple render menu', () => {
        const component = renderer.create(<HeaderMenu elements={this.elements} />).toJSON()

        expect(component).toMatchSnapshot()
    })
})
