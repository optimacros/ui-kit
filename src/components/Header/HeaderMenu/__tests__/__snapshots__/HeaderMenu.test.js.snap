// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderMenu simple render menu 1`] = `
<div
  className="Container"
>
  <ul
    className="Menu"
  >
    <li
      className="MenuItem"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      title="element 1"
    >
      <div
        className="Element ElementContainer"
        onClick={[Function]}
      >
        <div
          className="Element_Title"
        >
          element 1
        </div>
      </div>
    </li>
    <li
      className="MenuItem"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      title="element 2"
    >
      <div
        className="Element ElementContainer"
        onClick={[Function]}
      >
        <div
          className="Element_Title"
        >
          element 2
        </div>
      </div>
    </li>
    <li
      className="MenuItem"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      title="element 3"
    >
      <div
        className="Element ElementContainer"
        onClick={[Function]}
      >
        <div
          className="Element_Title"
        >
          element 3
        </div>
      </div>
    </li>
    <li
      className="MenuItem MenuItem__disabled"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      title="element 4"
    >
      <div
        className="Element ElementContainer"
        disabled={true}
        onClick={[Function]}
      >
        <div
          className="Element_Title"
        >
          element 4
        </div>
      </div>
    </li>
  </ul>
</div>
`;
